docs/stateless.md

```diff
@@ -0,0 +1,64 @@
+
+With the new feature of running the IRMA server in stateless mode, the session data gets decoupled from the IRMA server making the server itself stateless. The IRMA server still contains the logic when to switch between states but will retrieve and update the session data from and to a dedicated store. Hence, several IRMA servers can be run in parallel and connect to the same data store. Also, no data is lost when the IRMA server restarts or crashes.
+
+### Use cases
+* Scaling the IRMA server horizontally, i.e. running several IRMA severs in parallel.
+* Preventing data loss when restarting an IRMA server.
+* Preventing data loss in case of IRMA server crashes.
+
+### Stateless implementation using Redis
+The IRMA stateless mode uses Redis, an in-memory data structure store with on-disk persistence, to keep track of the session state. When a new session is started, two key/value pairs are added to the data store. The first key/value pair maps the `client token` (key) to a JSON blob which contains all stateful session data (value). Additionally, a second key/value pair is added to Redis which maps the `requestor token` (key) to the corresponding `client token` (value). This second key/value pair is needed so the session data can be retrieved not only when a `client token` is used for querying but also when a `requestor token` is used. In this later case, the `requestor token` will first be used to retrieve the `client token` with which the session data can be retrieved. The key/value pairs will expire eventually and will get removed by Redis automatically so the data store is not polluted with old sessions.  
+
+When a session is requested with a `client token` or a `requestor token`, a simple lock is introduced. The lock is another key/value pair with the `client token` as the key and a random identifier that is associated with the locking server as the value. The locked session can then be updated if necessary and will be unlocked again. This lock will also expire eventually and be removed automatically in case the unlocking for some reason did not succeed.
+
+> When running a single Redis instance only, this simple locking mechanism is sufficient. When running several Redis instances in parallel, edge cases may occur. With the current implementation, we recommend to run a single Redis instance only.
+
+## Running the IRMA server in stateless mode
+### Example
+You can start the IRMA server in stateless mode by passing the flag `--store-type redis` on the commandline. Additionally you need to provide a Redis server address and password. For test purposes you can overrule the need of a password by using the `--redis-allow-empty-password` flag. Make sure to use a secure Redis password in production though. Your Redis data store will contain sensitive data and must be password-protected.
```

Comment:
The first sentence mentions just the commandline flag, but all IRMA server commandline flags can also be passed as keys in a configuration file or environental variables (see [here](https://irma.app/docs/irma-server/#configuring)). Generally, when talking about IRMA server options I just call them "option", see the IRMA server page. (But, the example below is fine.)